cmd/sync/provider.go

```diff
@@ -0,0 +1,31 @@
+package main
+
+import "time"
+
+// CloudProvider is the interface to connect with any cloud provider.
+type CloudProvider interface {
```

Comment:
Let me know what you think about the following suggestions:

(1)
I think it is better to separate the non-cloud provider config from the cloud provider one so that the CloudProvider interface won't have to have `GetUpstreamsConfig()`, `GetAPIEndpoint()` and `GetSyncIntervalInSeconds()`. This way:
- The cloud provider interface becomes narrower
- Cloud provider implementations won't have to parse and validate the config they don't use

To implement that, we can have a struct like this:
```
type CommonConfig struct {
  PlusAPIEndpoint string
  GetSyncIntervalInSeconds time.Duration
  Upstreams []Upstreams
```

Such a struct can be parsed (and validated) from the configfile in `main.go`.  Later those parsed values can be used in the main loop.

(2)
Additionally, perhaps we can further simplify the interface by removing `Configure()` and `ValidateAndSaveConfig()` methods. I wonder if we really need those methods in the interface. Can we parse and validate  the config in `NewAWSClient()` function. (Note we can still have `configure` and `validate` functions, which NewAWSClient can invoke. However, those functions will be private helper function and won't be part of the CloudProviderInterface). If error happens, than `NewAWSClient()` can return an error.

 This way the interface will be easier to use (we won't have to remember the order `NewClient()` -> `ValidateAndSaveConfig()` -> `Configure()` -> `ready to use` and have `NewClient()` -> `ready to use` instead). In addition, this should make it easier to comprehend the interface because it has less methods.

Reply:
Those suggestions make sense thanks